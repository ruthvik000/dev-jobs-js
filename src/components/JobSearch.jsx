import React from 'react';
import Modal from "./Modal";

// Component to render search results.
function JobSearch(props) {
    const [resize, setResize] = React.useState(window.innerWidth);
    const [input, setInput] = React.useState('');
    const [checked, setChecked] = React.useState(false);
    const [locationInput, setLocationInput] = React.useState('');
    const [isOpen, setIsOpen] = React.useState(false);
    const handleClick = () => setChecked(!checked)
    const prevQueryParam = usePreviousQueryParamValue(props.queryParam);


    const onresize = function (e) {
        return setResize(e.target.innerWidth);
    }

    function onChangeInput(e) {
        setInput(e.target.value)
    }

    function onChangeLocation(e) {
        if (e.target.value === 'Default Location') {
            getLocation()
        } else {
            props.setLatitude(0);
            props.setLongitude(0);
        }
        setLocationInput(e.target.value)
    }

    function getLocation() {
        if (navigator.geolocation) {
            navigator.geolocation.getCurrentPosition(showPosition);
        }
    }

    function showPosition(position) {
        props.setLatitude(position.coords.latitude);
        props.setLongitude(position.coords.longitude);
    }

    function handleSearchBar() {
        let string = '';
        if (checked) {
            string += `&full_time=${checked}`
        }
        if (input && input.length > 0) {
            string += `&search=${input}`
        }

        if (props.latitude && props.longitude) {
            string += `&lat=${props.latitude}&long=${props.longitude}`
        } else if (locationInput && locationInput.length > 0) {
            string += `&location=${locationInput}`
        }

        if ((prevQueryParam.length > 0 && string === '') || string.length > 0) {
            props.setDisplayLoadMoreButton(false);
        }
        props.setPageNumber(1);
        props.setQueryParam(string);
    }

    function usePreviousQueryParamValue(value) {
        const ref = React.useRef();
        React.useEffect(() => {
            ref.current = value;
        })
        return ref.current;
    }

    React.useEffect(() => {
        window.addEventListener("resize", onresize);
    }, [resize])

    const searchPlaceHolder = resize < 768 ? 'Filter by Title...' : 'Filter by Title, companies, expertise...';

    let jobSearchInputField =
        <span>
            <i className="fa fa-search job-search" aria-hidden="true"/>
        <input className='search-input'
               type="searchInput"
               id="searchName"
               name="searchName"
               onChange={onChangeInput}
               value={input}
               placeholder={searchPlaceHolder}/>
        </span>

    let remainingFields =
        <span>
            <i className="fa fa-map-marker" aria-hidden="true"/>
            <input className='search-location'
                   type="searchLocation"
                   id="locationName"
                   list='default'
                   name="locationName"
                   value={locationInput}
                   onChange={onChangeLocation}
                   placeholder="Filter by Location..."/>
                   <datalist id="default">
                       <option value="Default Location"/>
                   </datalist>
        </span>
    let content = null;
    if (props.jobSearchHeader) {
        if (resize < 768) {
            content = <form className="search-form">
                {jobSearchInputField}
                <button style={{border: "none"}}
                        className='search-button-filter'
                        onClick={() => setIsOpen(true)}
                        type="button">
                    <i style={{backgroundColor: "none"}} className="fa fa-filter" aria-hidden="true"/>
                </button>
                <Modal open={isOpen} onClose={() => setIsOpen(false)}>
                    <div>
                        <input className="checkbox-field"
                               type="checkbox"
                               id="fullTime"
                               name="fullTime"
                               checked={checked}
                               onChange={handleClick}
                        />
                        <label htmlFor="fullTime" className='checkbox-label'>Full Time Only</label>
                    </div>
                    {remainingFields}
                </Modal>
                <button className="btn search-button-mobile" type="button" onClick={handleSearchBar}>
                    <i className="fa fa-search search-icon-mobile"/>
                </button>
            </form>
        } else {
            content =
                <form className="search-form">
                    {jobSearchInputField}
                    {remainingFields}
                    <span>
                        <input className="checkbox-field"
                               style={{display: 'inline-block', margin: '5px'}}
                               type="checkbox"
                               id="fullTime"
                               name="fullTime"
                               checked={checked}
                               onChange={handleClick}
                        />
                        <label htmlFor="fullTime" className='checkbox-label'>Full Time Only</label>
                        <input className='search-button' type="button" value="Search" onClick={handleSearchBar}/>
                    </span>
                </form>
        }
    }
    return (content);
}

export default (JobSearch)