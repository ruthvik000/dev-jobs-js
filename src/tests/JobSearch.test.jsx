import { shallow } from 'enzyme';
import JobSearch from "../components/JobSearch";

describe('Job search component', () => {
    const container = shallow(<JobSearch jobSearchHeader={true}/>);
    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('should have an search field', () => {
        expect(container.find('input[type="searchInput"]').length).toEqual(1);
    });

    it('should have an location field', () => {
        expect(container.find('input[type="searchLocation"]').length).toEqual(1);
    });

    it('should have check box and search field', () => {
        expect(container.find('input[type="checkbox"]').length).toEqual(1);
        expect(container.find('input[type="button"]').length).toEqual(1);
    });

    it('should have proper props for search job field', () => {
        expect(container.find('input[type="searchInput"]').props()).toEqual({
            className: 'search-input',
            id: 'searchName',
            name: 'searchName',
            onChange: expect.any(Function),
            placeholder: 'Filter by Title, companies, expertise...',
            value: '',
            type: 'searchInput',
        });
    });

    it('should have proper props for search location field', () => {
        expect(container.find('input[type="searchLocation"]').props()).toEqual({
            className: 'search-location',
            id: 'locationName',
            name: 'locationName',
            list: 'default',
            onChange: expect.any(Function),
            placeholder: 'Filter by Location...',
            value: '',
            type: 'searchLocation',
        });
    });

    it('should have proper props for check box', () => {
        expect(container.find('input[type="checkbox"]').props()).toEqual({
            className: 'checkbox-field',
            checked: false,
            id: 'fullTime',
            name: 'fullTime',
            style: {display: 'inline-block', margin: '5px'},
            onChange: expect.any(Function),
            type: 'checkbox',
        });
    });

    it('should have proper props for label field', () => {
        expect(container.find('label[className="checkbox-label"]').props()).toEqual({
            className: 'checkbox-label',
            children: 'Full Time Only',
            htmlFor: 'fullTime'
        });
    });

    it('should have proper props for search field', () => {
        expect(container.find('input[type="button"]').props()).toEqual({
            className: 'search-button',
            value: 'Search',
            onClick: expect.any(Function),
            type: 'button',
        });
    });
});