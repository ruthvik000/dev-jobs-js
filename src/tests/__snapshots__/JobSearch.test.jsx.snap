// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job search component should match the snapshot 1`] = `"<form class=\\"search-form\\"><span><i class=\\"fa fa-search job-search\\" aria-hidden=\\"true\\"></i><input type=\\"searchInput\\" class=\\"search-input\\" id=\\"searchName\\" name=\\"searchName\\" value=\\"\\" placeholder=\\"Filter by Title, companies, expertise...\\"/></span><span><i class=\\"fa fa-map-marker\\" aria-hidden=\\"true\\"></i><input type=\\"searchLocation\\" class=\\"search-location\\" id=\\"locationName\\" list=\\"default\\" name=\\"locationName\\" value=\\"\\" placeholder=\\"Filter by Location...\\"/><datalist id=\\"default\\"><option value=\\"Default Location\\"></option></datalist></span><span><input type=\\"checkbox\\" class=\\"checkbox-field\\" style=\\"display:inline-block;margin:5px\\" id=\\"fullTime\\" name=\\"fullTime\\"/><label for=\\"fullTime\\" class=\\"checkbox-label\\">Full Time Only</label><input type=\\"button\\" class=\\"search-button\\" value=\\"Search\\"/></span></form>"`;
